zallet/src/components/json_rpc/methods/view_transaction.rs

```diff
@@ -472,9 +676,141 @@ pub(crate) fn call(wallet: &DbConnection, txid_str: &str) -> Response {
         }
     }
 
+    let wallet_tx_info = WalletTxInfo::fetch(wallet, &chain, &tx, chain_height)
+        .await
+        .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?;
+
+    let fee = tx
+        .fee_paid(|prevout| Ok::<_, BalanceError>(transparent_input_values.get(prevout).copied()))
+        // This should never occur, as a transaction that violated balance would be
+        // rejected by the backing full node.
+        .map_err(|e| LegacyCode::Database.with_message(e.to_string()))?;
+
     Ok(Transaction {
-        txid: txid_str.into(),
+        txid: txid_str.to_ascii_lowercase(),
+        status: wallet_tx_info.status,
+        confirmations: wallet_tx_info.confirmations,
+        blockhash: wallet_tx_info.blockhash,
+        blockindex: wallet_tx_info.blockindex,
+        blocktime: wallet_tx_info.blocktime,
+        version: tx.version().header() & 0x7FFFFFFF,
+        expiryheight: wallet_tx_info.expiryheight,
+        fee: fee.map(value_from_zatoshis),
+        generated: wallet_tx_info.generated,
         spends,
         outputs,
     })
 }
+
+struct WalletTxInfo {
+    status: &'static str,
+    confirmations: i32,
+    generated: Option<bool>,
+    blockhash: Option<String>,
+    blockindex: Option<u16>,
+    blocktime: Option<u64>,
+    expiryheight: u64,
+}
+
+impl WalletTxInfo {
+    /// Logic adapted from `WalletTxToJSON` in `zcashd`, to match the semantics of the `gettransaction` fields.
+    async fn fetch(
+        wallet: &DbConnection,
+        chain: &FetchServiceSubscriber,
+        tx: &zcash_primitives::transaction::Transaction,
+        chain_height: BlockHeight,
+    ) -> Result<Self, SqliteClientError> {
+        let mined_height = wallet.get_tx_height(tx.txid())?;
+
+        let confirmations = {
+            match mined_height {
+                Some(mined_height) => (chain_height + 1 - mined_height) as i32,
```

Comment:
We should use i64 here; even though we "know" the cast from block height to i32 is safe, using `i64::from(...)` would be cleaner.